[VideoFrame] Avoid internal MT in Accelerate

mstyura opened the pull request:
#### ecb227c4bef99549938b69d6eaaa85357c9444e0
<pre>
[VideoFrame] Avoid internal MT in Accelerate

<a href="https://bugs.webkit.org/show_bug.cgi?id=299819">https://bugs.webkit.org/show_bug.cgi?id=299819</a>

Reviewed by Darin Adler.

VideoFrame construction in Safari shows a ~10x slowdown compared to CI builds,
seemingly due to a different allocator selected at runtime. Safari uses the
secure XZM allocator, while CI builds use the default allocator. A working
hypothesis is that because the secure allocator lazily zeroes memory on first
access, and Accelerate by default tiles the image and performs a multithreaded
copy, it introduces contention when multiple threads touch newly mapped pages
that must be zeroed. Since the secure allocator cannot be enabled predictably
for testing, a speculative fix is to disable multithreaded copy in Accelerate
by disallowing it from tiling the image via `kvImageDoNotTile`, which is
expected to eliminate contention on zeroing.

* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createRGBA):
(WebCore::VideoFrame::createBGRA):

Canonical link: <a href="https://commits.webkit.org/300781@main">https://commits.webkit.org/300781@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7e611a6579166a3c2c2cd5a8d9eac7cc1151260

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130597 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75965 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125692 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44254 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52124 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94187 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62490 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110765 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74763 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34210 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28925 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74077 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104981 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29148 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133290 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50768 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38660 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102638 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51142 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106984 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102464 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26056 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47811 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26066 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47617 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50621 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56385 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50095 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53441 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mstyura commented:
The benchmark from [PR#48205](https://github.com/WebKit/WebKit/pull/48205) was upload to GitHub pages: https://mstyura.github.io/webkit-issues/video-frame-construction/ to make it easier for anyone to inspect performance.

darinadler reviewed (approved):
You’ll need to fix the change log and remove `No new tests (OOPS!).`

I hate to make things more wordy, but I also think we need a comment explaining why we pass "do not tile".

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b7e611a6579166a3c2c2cd5a8d9eac7cc1151260. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51522

mstyura committed:
[VideoFrame] Avoid internal MT in Accelerate

https://bugs.webkit.org/show_bug.cgi?id=299819

Reviewed by Darin Adler.

VideoFrame construction in Safari shows a ~10x slowdown compared to CI builds,
seemingly due to a different allocator selected at runtime. Safari uses the
secure XZM allocator, while CI builds use the default allocator. A working
hypothesis is that because the secure allocator lazily zeroes memory on first
access, and Accelerate by default tiles the image and performs a multithreaded
copy, it introduces contention when multiple threads touch newly mapped pages
that must be zeroed. Since the secure allocator cannot be enabled predictably
for testing, a speculative fix is to disable multithreaded copy in Accelerate
by disallowing it from tiling the image via `kvImageDoNotTile`, which is
expected to eliminate contention on zeroing.

* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createRGBA):
(WebCore::VideoFrame::createBGRA):

Canonical link: https://commits.webkit.org/300781@main
